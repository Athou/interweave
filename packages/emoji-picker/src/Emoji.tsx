/**
 * @copyright   2016-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import React from 'react';
import PropTypes from 'prop-types';
import EmojiCharacter, {
  CanonicalEmoji,
  EmojiShape,
  Path,
  PathShape,
  Source,
  SourceShape,
} from 'interweave-emoji';
import withContext, { ContextProps } from './Context';
import { ContextShape } from './shapes';

export interface EmojiProps {
  active: boolean;
  emoji: CanonicalEmoji;
  emojiPadding: number;
  emojiPath: Path;
  emojiSize: number;
  emojiSource: Source;
  onEnter: (emoji: CanonicalEmoji, event: React.MouseEvent<HTMLButtonElement>) => void;
  onLeave: (emoji: CanonicalEmoji, event: React.MouseEvent<HTMLButtonElement>) => void;
  onSelect: (emoji: CanonicalEmoji, event: React.MouseEvent<HTMLButtonElement>) => void;
  showImage: boolean;
}

export interface EmojiState {
  active: boolean;
}

export class Emoji extends React.PureComponent<EmojiProps & ContextProps, EmojiState> {
  static propTypes = {
    active: PropTypes.bool.isRequired,
    context: ContextShape.isRequired,
    emoji: EmojiShape.isRequired,
    emojiPadding: PropTypes.number.isRequired,
    emojiPath: PathShape.isRequired,
    emojiSize: PropTypes.number.isRequired,
    emojiSource: SourceShape.isRequired,
    onEnter: PropTypes.func.isRequired,
    onLeave: PropTypes.func.isRequired,
    onSelect: PropTypes.func.isRequired,
    showImage: PropTypes.bool.isRequired,
  };

  state = {
    active: this.props.active,
  };

  componentDidUpdate(prevProps: EmojiProps) {
    if (this.props.active !== prevProps.active) {
      this.setState({
        active: this.props.active,
      });
    }
  }

  /**
   * Triggered when the emoji is clicked.
   */
  private handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();

    this.props.onSelect(this.props.emoji, event);
  };

  /**
   * Triggered when the emoji is hovered.
   */
  private handleEnter = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();

    this.setState({
      active: true,
    });

    this.props.onEnter(this.props.emoji, event);
  };

  /**
   * Triggered when the emoji is no longer hovered.
   */
  private handleLeave = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();

    this.setState({
      active: false,
    });

    this.props.onLeave(this.props.emoji, event);
  };

  render() {
    const {
      context: { classNames },
      emoji,
      emojiPadding,
      emojiPath,
      emojiSize,
      emojiSource,
      showImage,
    } = this.props;
    const { active } = this.state;
    const dimension = emojiPadding + emojiPadding + emojiSize;
    const className = [classNames.emoji];

    if (active) {
      className.push(classNames.emojiActive);
    }

    return (
      <button
        aria-label={emoji.annotation}
        key={emoji.hexcode}
        className={className.join(' ')}
        style={{
          height: dimension,
          padding: emojiPadding,
          width: dimension,
        }}
        type="button"
        onClick={this.handleClick}
        onMouseEnter={this.handleEnter}
        onMouseLeave={this.handleLeave}
      >
        {showImage ? (
          <EmojiCharacter
            emojiPath={emojiPath}
            emojiSize={emojiSize}
            emojiSource={emojiSource}
            hexcode={emoji.hexcode}
          />
        ) : (
          <div
            style={{
              height: emojiSize,
              overflow: 'hidden',
              visibility: 'hidden',
              width: emojiSize,
            }}
          >
            &nbsp;
          </div>
        )}
      </button>
    );
  }
}

export default withContext(Emoji);
