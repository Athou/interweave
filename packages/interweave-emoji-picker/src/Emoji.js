/**
 * @copyright   2016, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import React from 'react';
import PropTypes from 'prop-types';
import EmojiCharacter from 'interweave/lib/components/Emoji';
import { EmojiShape, EmojiPathShape } from './shapes';

import type { Emoji, EmojiPath } from './types';

type EmojiProps = {
  emoji: Emoji,
  emojiPath: EmojiPath,
  emojiSize: number,
  onEnter: (emoji: Emoji) => void,
  onLeave: (emoji: Emoji) => void,
  onSelect: (emoji: Emoji) => void,
};

type EmojiState = {
  active: boolean,
};

export default class EmojiButton extends React.PureComponent<EmojiProps, EmojiState> {
  static propTypes = {
    emoji: EmojiShape.isRequired,
    emojiPath: EmojiPathShape.isRequired,
    emojiSize: PropTypes.number.isRequired,
    onEnter: PropTypes.func.isRequired,
    onLeave: PropTypes.func.isRequired,
    onSelect: PropTypes.func.isRequired,
  };

  state = {
    active: false,
  };

  handleEnter = (e: SyntheticMouseEvent<*>) => {
    e.stopPropagation();

    this.setState({
      active: true,
    });

    this.props.onEnter(this.props.emoji);
  };

  handleLeave = (e: SyntheticMouseEvent<*>) => {
    e.stopPropagation();

    this.setState({
      active: false,
    });

    this.props.onLeave(this.props.emoji);
  };

  handleSelect = (e: SyntheticMouseEvent<*>) => {
    e.stopPropagation();

    this.props.onSelect(this.props.emoji);
  };

  render() {
    const { emoji, emojiPath, emojiSize } = this.props;
    let className = 'iep__emoji';

    if (this.state.active) {
      className += ' iep__emoji--active';
    }

    return (
      <button
        key={emoji.hexcode}
        type="button"
        className={className}
        onClick={this.handleSelect}
        onMouseEnter={this.handleEnter}
        onMouseLeave={this.handleLeave}
      >
        <EmojiCharacter
          unicode={emoji.emoji || emoji.text}
          emojiPath={emojiPath}
          emojiSize={emojiSize}
        />
      </button>
    );
  }
}
