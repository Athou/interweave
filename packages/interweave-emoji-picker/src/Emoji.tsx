/**
 * @copyright   2016, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import React from 'react';
import PropTypes from 'prop-types';
import EmojiCharacter, {
  CanonicalEmoji,
  EmojiShape,
  EmojiPath,
  EmojiPathShape,
  EmojiSource,
  EmojiSourceShape,
} from 'interweave-emoji';

export interface EmojiProps {
  active: boolean;
  emoji: CanonicalEmoji;
  emojiPadding: number;
  emojiPath: EmojiPath;
  emojiSize: number;
  emojiSource: EmojiSource;
  onEnter: (emoji: CanonicalEmoji, e: any) => void;
  onLeave: (emoji: CanonicalEmoji, e: any) => void;
  onSelect: (emoji: CanonicalEmoji, e: any) => void;
  showImage: boolean;
}

export interface EmojiState {
  active: boolean;
}

export default class EmojiButton extends React.PureComponent<EmojiProps, EmojiState> {
  static contextTypes = {
    classNames: PropTypes.objectOf(PropTypes.string),
  };

  static propTypes = {
    active: PropTypes.bool.isRequired,
    emoji: EmojiShape.isRequired,
    emojiPadding: PropTypes.number.isRequired,
    emojiPath: EmojiPathShape.isRequired,
    emojiSize: PropTypes.number.isRequired,
    emojiSource: EmojiSourceShape.isRequired,
    onEnter: PropTypes.func.isRequired,
    onLeave: PropTypes.func.isRequired,
    onSelect: PropTypes.func.isRequired,
    showImage: PropTypes.bool.isRequired,
  };

  state = {
    active: this.props.active,
  };

  componentDidUpdate(prevProps: EmojiProps) {
    if (this.props.active !== prevProps.active) {
      this.setState({
        active: this.props.active,
      });
    }
  }

  /**
   * Triggered when the emoji is clicked.
   */
  private handleClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.stopPropagation();

    this.props.onSelect(this.props.emoji, e);
  };

  /**
   * Triggered when the emoji is hovered.
   */
  private handleEnter = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.stopPropagation();

    this.setState({
      active: true,
    });

    this.props.onEnter(this.props.emoji, e);
  };

  /**
   * Triggered when the emoji is no longer hovered.
   */
  private handleLeave = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.stopPropagation();

    this.setState({
      active: false,
    });

    this.props.onLeave(this.props.emoji, e);
  };

  render() {
    const { emoji, emojiPadding, emojiPath, emojiSize, emojiSource, showImage } = this.props;
    const { classNames } = this.context;
    const { active } = this.state;
    const dimension = emojiPadding + emojiPadding + emojiSize;
    const className = [classNames.emoji];

    if (active) {
      className.push(classNames.emojiActive);
    }

    return (
      <button
        aria-label={emoji.annotation}
        key={emoji.hexcode}
        className={className.join(' ')}
        style={{
          height: dimension,
          padding: emojiPadding,
          width: dimension,
        }}
        type="button"
        onClick={this.handleClick}
        onMouseEnter={this.handleEnter}
        onMouseLeave={this.handleLeave}
      >
        {showImage ? (
          <EmojiCharacter
            emojiPath={emojiPath}
            emojiSize={emojiSize}
            emojiSource={emojiSource}
            hexcode={emoji.hexcode}
          />
        ) : (
          <div
            style={{
              height: emojiSize,
              overflow: 'hidden',
              visibility: 'hidden',
              width: emojiSize,
            }}
          >
            &nbsp;
          </div>
        )}
      </button>
    );
  }
}
