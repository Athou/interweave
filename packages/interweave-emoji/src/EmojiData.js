/**
 * @copyright   2016, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

/* eslint-disable no-param-reassign */

import {
  fromCodepointToUnicode,
  fromHexcodeToCodepoint,
  generateEmoticonPermutations,
} from 'emojibase';
import { TEXT, EMOTICON_OPTIONS } from 'emojibase/lib/constants';

import type { Emoji } from 'emojibase';

const instances: { [locale: string]: EmojiData } = {}; // eslint-disable-line

export default class EmojiData {
  EMOJIS: { [hexcode: string]: Emoji } = {};
  EMOTICON_TO_HEXCODE: { [emoticon: string]: string } = {};
  SHORTCODE_TO_HEXCODE: { [shortcode: string]: string } = {};
  UNICODE_TO_HEXCODE: { [unicode: string]: string } = {};

  data: Emoji[] = [];
  flatData: Emoji[] = [];
  locale: string = 'en';

  constructor(locale: string) {
    this.locale = locale;
  }

  /**
   * Return or create a singleton instance per locale.
   */
  static getInstance(locale: string) {
    if (!instances[locale]) {
      instances[locale] = new EmojiData(locale);
    }

    return instances[locale];
  }

  /**
   * Return dataset as a list.
   */
  getData(): Emoji[] {
    return this.data;
  }

  /**
   * Return dataset as a flattened list.
   */
  getFlatData(): Emoji[] {
    return this.flatData;
  }

  /**
   * Package the emoji object with additional data,
   * while also extracting and partitioning relevant information.
   */
  packageEmoji(emoji: Object): Emoji {
    const { emoticon, hexcode, shortcodes } = emoji;

    // Make our lives easier
    if (!emoji.unicode) {
      emoji.unicode = (emoji.text && emoji.type === TEXT) ? emoji.text : emoji.emoji;
    }

    // Canonicalize the shortcodes for easy reuse
    emoji.canonical_shortcodes = shortcodes.map(code => `:${code}:`);
    emoji.primary_shortcode = emoji.canonical_shortcodes[0]; // eslint-disable-line

    // Support all shortcodes
    emoji.canonical_shortcodes.forEach((shortcode) => {
      this.SHORTCODE_TO_HEXCODE[shortcode] = hexcode;
    });

    // Support all emoticons
    if (emoticon) {
      generateEmoticonPermutations(emoticon, EMOTICON_OPTIONS[emoticon]).forEach((emo) => {
        this.EMOTICON_TO_HEXCODE[emo] = hexcode;
      });
    }

    // Support all presentations (even no variation selectors)
    this.UNICODE_TO_HEXCODE[fromCodepointToUnicode(fromHexcodeToCodepoint(hexcode))] = hexcode;

    if (emoji.emoji) {
      this.UNICODE_TO_HEXCODE[emoji.emoji] = hexcode;
    }

    if (emoji.text) {
      this.UNICODE_TO_HEXCODE[emoji.text] = hexcode;
    }

    // Map each emoji
    this.EMOJIS[hexcode] = emoji;

    return emoji;
  }

  /**
   * Parse and generate emoji datasets.
   */
  parseEmojiData(data: Emoji[]): Emoji[] {
    data.forEach((emoji) => {
      const packagedEmoji = this.packageEmoji(emoji);

      this.data.push(packagedEmoji);
      this.flatData.push(packagedEmoji);

      // Flatten and package skins as well
      if (packagedEmoji.skins) {
        packagedEmoji.skins.forEach((skin) => {
          this.flatData.push(this.packageEmoji(skin));
        });
      }
    });

    return this.data;
  }
}
