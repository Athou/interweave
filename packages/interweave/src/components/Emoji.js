/**
 * @copyright   2016, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

/* eslint-disable complexity, no-param-reassign */

import React from 'react';
import PropTypes from 'prop-types';
import { fromHexcodeToCodepoint } from 'emojibase';
import {
  EMOJIS,
  EMOTICON_TO_UNICODE,
  SHORTCODE_TO_UNICODE,
  UNICODE_TO_SHORTCODES,
} from '../data/emoji';

import type { EmojiProps } from '../types';

const LARGE_MULTIPLIER: number = 3;

export default class Emoji extends React.PureComponent<EmojiProps> {
  static propTypes = {
    emojiLargeSize: PropTypes.number,
    emojiPath: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.func,
    ]),
    emojiSize: PropTypes.number,
    emoticon: PropTypes.string,
    enlargeEmoji: PropTypes.bool,
    shortcode: PropTypes.string,
    unicode: PropTypes.string,
  };

  static defaultProps = {
    emojiLargeSize: 0,
    emojiPath: '{{hexcode}}',
    emojiSize: 0,
    emoticon: '',
    enlargeEmoji: false,
    shortcode: '',
    unicode: '',
  };

  render() {
    const {
      emojiLargeSize,
      emojiPath,
      emojiSize,
      emoticon,
      enlargeEmoji,
      shortcode,
    } = this.props;
    let { unicode } = this.props;

    if (__DEV__) {
      if (!emoticon && !shortcode && !unicode) {
        throw new Error(
          'Emoji component requires a `unicode` character, `emoticon`, or a `shortcode`.',
        );
      }
    }

    // Retrieve applicable unicode character
    if (!unicode && shortcode) {
      unicode = SHORTCODE_TO_UNICODE[shortcode];
    }

    if (!unicode && emoticon) {
      unicode = EMOTICON_TO_UNICODE[emoticon];
    }

    // Return the invalid value instead of erroring
    if (!unicode || !EMOJIS[unicode]) {
      return (
        <span>{unicode || emoticon || shortcode}</span>
      );
    }

    const emoji = EMOJIS[unicode];
    const className = ['interweave__emoji'];
    const styles = {};

    // Only apply styles if a size is defined
    if (emojiSize) {
      styles.display = 'inline-block';
      styles.verticalAlign = 'middle';
      styles.width = `${emojiSize}em`;
    }

    // Handle large styles
    if (enlargeEmoji) {
      className.push('interweave__emoji--large');

      if (emojiLargeSize) {
        styles.width = `${emojiLargeSize}em`;
      } else if (emojiSize) {
        styles.width = `${emojiSize * LARGE_MULTIPLIER}em`;
      }
    }

    // Determine the path
    let path = emojiPath || '{{hexcode}}';

    if (typeof path === 'function') {
      path = path(emoji.hexcode, enlargeEmoji, emojiSize, emojiLargeSize);
    } else {
      path = path.replace('{{hexcode}}', emoji.hexcode);
    }

    // http://git.emojione.com/demos/latest/sprites-png.html
    // http://git.emojione.com/demos/latest/sprites-svg.html
    // https://css-tricks.com/using-svg/
    return (
      <img
        src={path}
        alt={unicode}
        title={emoji.annotation || ''}
        style={styles}
        className={className.join(' ')}
        aria-label={emoji.annotation || ''}
        data-emoticon={emoji.emoticon || ''}
        data-unicode={unicode}
        data-hexcode={emoji.hexcode}
        data-shortcodes={UNICODE_TO_SHORTCODES[unicode].join(', ')}
        data-codepoint={fromHexcodeToCodepoint(emoji.hexcode).join('-')}
      />
    );
  }
}
