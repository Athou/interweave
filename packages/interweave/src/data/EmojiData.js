/**
 * @copyright   2016, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

/* eslint-disable no-param-reassign */

import { generateEmoticonPermutations } from 'emojibase';
import { TEXT, EMOTICON_OPTIONS } from 'emojibase/lib/constants';

import type { Emoji } from 'emojibase';

const instances: { [locale: string]: EmojiData } = {}; // eslint-disable-line

export default class EmojiData {
  EMOJIS: { [unicode: string]: Emoji } = {};
  EMOTICON_TO_UNICODE: { [emoticon: string]: string } = {};
  SHORTCODE_TO_UNICODE: { [shortcode: string]: string } = {};
  UNICODE_TO_SHORTCODES: { [unicode: string]: string[] } = {};

  data: Emoji[] = [];
  flatData: Emoji[] = [];
  locale: string = 'en';

  constructor(locale: string) {
    this.locale = locale;
  }

  static getInstance(locale: string) {
    if (!instances[locale]) {
      instances[locale] = new EmojiData(locale);
    }

    return instances[locale];
  }

  getData(): Emoji[] {
    return this.data;
  }

  getFlatData(): Emoji[] {
    return this.flatData;
  }

  packageEmoji(emoji: Object): Emoji {
    const { emoticon, shortcodes } = emoji;

    // Only support the default presentation
    const unicode = (emoji.text && emoji.type === TEXT) ? emoji.text : emoji.emoji;
    emoji.unicode = unicode;

    // Canonicalize the shortcodes for easy reuse
    emoji.canonical_shortcodes = shortcodes.map(code => `:${code}:`);
    emoji.primary_shortcode = emoji.canonical_shortcodes[0]; // eslint-disable-line

    // Support all shortcodes
    this.UNICODE_TO_SHORTCODES[unicode] = emoji.canonical_shortcodes.map((shortcode) => {
      this.SHORTCODE_TO_UNICODE[shortcode] = unicode;

      return shortcode;
    });

    // Support all emoticons
    if (emoticon) {
      generateEmoticonPermutations(emoticon, EMOTICON_OPTIONS[emoticon]).forEach((emo) => {
        this.EMOTICON_TO_UNICODE[emo] = unicode;
      });
    }

    // Map each emoji by unicode
    this.EMOJIS[unicode] = emoji;

    return emoji;
  }

  parseEmojiData(data: Emoji[]): Emoji[] {
    data.forEach((emoji) => {
      const packagedEmoji = this.packageEmoji(emoji);

      this.data.push(packagedEmoji);
      this.flatData.push(packagedEmoji);

      // Flatten and package skins as well
      if (packagedEmoji.skins) {
        packagedEmoji.skins.forEach((skin) => {
          this.flatData.push(this.packageEmoji(skin));
        });
      }
    });

    return this.data;
  }
}
