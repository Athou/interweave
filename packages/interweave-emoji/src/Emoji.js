/**
 * @copyright   2016, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

/* eslint-disable complexity, no-param-reassign */

import React from 'react';
import PropTypes from 'prop-types';
import EmojiData from './EmojiData';
import {
  EmojiPathShape,
  EmojiSizeShape,
  EmojiSourceShape,
} from './shapes';

type EmojiSize = string | number;

type EmojiProps = {
  emojiLargeSize: EmojiSize,
  emojiPath: string |
    (hexcode: string, enlarged: boolean, smallSize: EmojiSize, largeSize: EmojiSize) => string,
  emojiSize: EmojiSize,
  emojiSource: {
    compact: boolean,
    locale: string,
    version: string,
  },
  emoticon: string,
  enlargeEmoji: boolean,
  hexcode: string,
  renderUnicode: boolean,
  shortcode: string,
  unicode: string,
};

export default class Emoji extends React.PureComponent<EmojiProps> {
  static propTypes = {
    emojiLargeSize: EmojiSizeShape,
    emojiPath: EmojiPathShape,
    emojiSize: EmojiSizeShape,
    emojiSource: EmojiSourceShape.isRequired,
    emoticon: PropTypes.string,
    enlargeEmoji: PropTypes.bool,
    hexcode: PropTypes.string,
    renderUnicode: PropTypes.bool,
    shortcode: PropTypes.string,
    unicode: PropTypes.string,
  };

  static defaultProps = {
    emojiLargeSize: '3em',
    emojiPath: '{{hexcode}}',
    emojiSize: '1em',
    emoticon: '',
    enlargeEmoji: false,
    hexcode: '',
    renderUnicode: false,
    shortcode: '',
    unicode: '',
  };

  render() {
    const data = EmojiData.getInstance(this.props.emojiSource.locale);
    const {
      emojiLargeSize,
      emojiPath,
      emojiSize,
      emoticon,
      enlargeEmoji,
      renderUnicode,
      shortcode,
      unicode,
    } = this.props;
    let { hexcode } = this.props;

    if (__DEV__) {
      if (!emoticon && !shortcode && !unicode && !hexcode) {
        throw new Error(
          'Emoji component requires a `unicode` character, `emoticon`, `hexcode`, or a `shortcode`.',
        );
      }
    }

    // Retrieve applicable unicode character
    if (!hexcode && shortcode) {
      hexcode = data.SHORTCODE_TO_HEXCODE[shortcode];
    }

    if (!hexcode && emoticon) {
      hexcode = data.EMOTICON_TO_HEXCODE[emoticon];
    }

    if (!hexcode && unicode) {
      hexcode = data.UNICODE_TO_HEXCODE[unicode];
    }

    // Return the invalid value instead of erroring
    if (!hexcode || !data.EMOJIS[hexcode]) {
      return (
        <span>{unicode || emoticon || shortcode || hexcode}</span>
      );
    }

    const emoji = data.EMOJIS[hexcode];

    if (renderUnicode) {
      return (
        <span>{emoji.unicode}</span>
      );
    }

    const className = ['interweave__emoji'];
    const styles: Object = {
      display: 'inline-block',
      verticalAlign: 'middle',
    };

    // Only apply styles if a size is defined
    if (emojiSize) {
      styles.width = emojiSize;
      styles.height = emojiSize;
    }

    // Handle large styles
    if (enlargeEmoji) {
      className.push('interweave__emoji--large');

      if (emojiLargeSize) {
        styles.width = emojiLargeSize;
        styles.height = emojiLargeSize;
      }
    }

    // Determine the path
    let path = emojiPath || '{{hexcode}}';

    if (typeof path === 'function') {
      path = path(emoji.hexcode, enlargeEmoji, emojiSize, emojiLargeSize);
    } else {
      path = path.replace('{{hexcode}}', emoji.hexcode);
    }

    // http://git.emojione.com/demos/latest/sprites-png.html
    // http://git.emojione.com/demos/latest/sprites-svg.html
    // https://css-tricks.com/using-svg/
    return (
      <img
        src={path}
        alt={emoji.unicode}
        title={emoji.annotation || ''}
        style={styles}
        className={className.join(' ')}
        aria-label={emoji.annotation || ''}
        data-emoticon={emoji.emoticon || ''}
        data-unicode={emoji.unicode}
        data-hexcode={emoji.hexcode}
        /* $FlowIgnore */
        data-shortcodes={emoji.canonical_shortcodes.join(', ')}
      />
    );
  }
}
