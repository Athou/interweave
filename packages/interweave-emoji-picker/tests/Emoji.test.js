import React from 'react';
import { shallow } from 'enzyme';
import EmojiCharacter from '../../interweave-emoji/src/Emoji';
import EmojiData from '../../interweave-emoji/src/EmojiData';
import Emoji from '../src/Emoji';
import { SOURCE_PROP } from '../../../tests/mocks';

describe('Emoji', () => {
  const props = {
    active: false,
    emoji: {},
    emojiPadding: 5,
    emojiPath: '{{hexcode}}',
    emojiSize: 22,
    emojiSource: SOURCE_PROP,
    showImage: true,
    onEnter() {},
    onLeave() {},
    onSelect() {},
  };

  const event = {
    preventDefault() {},
    stopPropagation() {},
  };

  beforeEach(() => {
    props.emoji = EmojiData.getInstance('en').SHORTCODE_TO_UNICODE[':cat:'];
  });

  it('renders an emoji', () => {
    const wrapper = shallow(<Emoji {...props} />);
    const char = wrapper.find(EmojiCharacter);

    expect(char).toHaveLengthOf(1);
    expect(char.prop('unicode')).toBe(props.emoji.unicode);
  });

  it('renders a placeholder if images are hidden', () => {
    const wrapper = shallow(<Emoji {...props} showImage={false} />);

    expect(wrapper.find(EmojiCharacter)).toHaveLengthOf(0);
    expect(wrapper.find('div').prop('style')).toEqual({
      width: 32,
      height: 32,
      overflow: 'hidden',
      visibility: 'hidden',
    });
  });

  it('updates the active state', () => {
    const wrapper = shallow(<Emoji {...props} />);

    expect(wrapper.prop('className')).toBe('interweave-picker__emoji');

    wrapper.setProps({
      active: true,
    });

    expect(wrapper.prop('className'))
      .toBe('interweave-picker__emoji interweave-picker__emoji--active');
  });

  it('sets correct sizes', () => {
    const wrapper = shallow(<Emoji {...props} />);

    expect(wrapper.prop('style')).toEqual({
      width: 32,
      height: 32,
      padding: 5,
    });
  });

  it('triggers `onSelect` when clicking', () => {
    const spy = jest.fn();
    const wrapper = shallow(<Emoji {...props} onSelect={spy} />);

    wrapper.simulate('click', event);

    expect(spy).toBeCalledWith(props.emoji, event);
  });

  it('triggers `onEnter` when entering node', () => {
    const spy = jest.fn();
    const wrapper = shallow(<Emoji {...props} onEnter={spy} />);

    wrapper.simulate('mouseenter', event);

    expect(spy).toBeCalledWith(props.emoji, event);
    expect(wrapper.state('active')).toBe(true);
  });

  it('triggers `onLeave` when leaving node', () => {
    const spy = jest.fn();
    const wrapper = shallow(<Emoji {...props} onLeave={spy} />);

    wrapper.simulate('mouseenter', event);

    expect(spy).toBeCalledWith(props.emoji, event);
    expect(wrapper.state('active')).toBe(false);
  });
});
