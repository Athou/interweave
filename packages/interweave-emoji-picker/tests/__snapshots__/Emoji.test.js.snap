// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emoji renders an emoji 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Emoji
    active={false}
    context={
      Object {
        "classNames": Object {
          "emoji": "interweave-picker__emoji",
          "emojiActive": "interweave-picker__emoji--active",
          "emojis": "interweave-picker__emojis",
          "emojisBody": "interweave-picker__emojis-body",
          "emojisContainer": "interweave-picker__emojis-container",
          "emojisHeader": "interweave-picker__emojis-header",
          "emojisRow": "interweave-picker__emojis-row",
          "emojisSection": "interweave-picker__emojis-section",
          "group": "interweave-picker__group",
          "groupActive": "interweave-picker__group--active",
          "groups": "interweave-picker__groups",
          "groupsList": "interweave-picker__groups-list",
          "noPreview": "interweave-picker__no-preview",
          "noResults": "interweave-picker__no-results",
          "picker": "interweave-picker__picker",
          "pickerVirtual": "interweave-picker__picker--virtual",
          "preview": "interweave-picker__preview",
          "previewContent": "interweave-picker__preview-content",
          "previewEmoji": "interweave-picker__preview-emoji",
          "previewSubtitle": "interweave-picker__preview-subtitle",
          "previewTitle": "interweave-picker__preview-title",
          "search": "interweave-picker__search",
          "searchInput": "interweave-picker__search-input",
          "skinTone": "interweave-picker__skin-tone",
          "skinToneActive": "interweave-picker__skin-tone--active",
          "skinTones": "interweave-picker__skin-tones",
        },
        "messages": Object {
          "activities": "Activities",
          "animals-nature": "Animals & Nature",
          "flags": "Flags",
          "food-drink": "Food & Drink",
          "frequently-used": "Frequently Used",
          "noPreview": "No emoji to display",
          "noResults": "No results",
          "none": "All Emojis",
          "objects": "Objects",
          "recently-used": "Recently Used",
          "search": "Search",
          "search-results": "Search Results",
          "searchA11y": "Search for emojis by keyword",
          "smileys-people": "Smileys & People",
          "symbols": "Symbols",
          "travel-places": "Travel & Places",
        },
      }
    }
    emoji={
      Object {
        "annotation": "cat",
        "canonical_shortcodes": Array [
          ":cat:",
        ],
        "emoji": "🐈️",
        "group": 1,
        "hexcode": "1F408",
        "name": "CAT",
        "order": 2079,
        "primary_shortcode": ":cat:",
        "shortcodes": Array [
          "cat",
        ],
        "skins": undefined,
        "subgroup": 18,
        "tags": Array [
          "pet",
        ],
        "text": "🐈︎",
        "type": 1,
        "unicode": "🐈️",
        "version": 1,
      }
    }
    emojiPadding={5}
    emojiPath="{{hexcode}}"
    emojiSize={22}
    emojiSource={
      Object {
        "compact": false,
        "locale": "en",
        "version": "latest",
      }
    }
    onEnter={[Function]}
    onLeave={[Function]}
    onSelect={[Function]}
    showImage={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "1F408",
    "nodeType": "host",
    "props": Object {
      "aria-label": "cat",
      "children": <Emoji
        emojiLargeSize="3em"
        emojiPath="{{hexcode}}"
        emojiSize={22}
        emojiSource={
          Object {
            "compact": false,
            "locale": "en",
            "version": "latest",
          }
        }
        emoticon=""
        enlargeEmoji={false}
        hexcode="1F408"
        renderUnicode={false}
        shortcode=""
        unicode=""
      />,
      "className": "interweave-picker__emoji",
      "onClick": [Function],
      "onMouseEnter": [Function],
      "onMouseLeave": [Function],
      "style": Object {
        "height": 32,
        "padding": 5,
        "width": 32,
      },
      "type": "button",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "emojiLargeSize": "3em",
        "emojiPath": "{{hexcode}}",
        "emojiSize": 22,
        "emojiSource": Object {
          "compact": false,
          "locale": "en",
          "version": "latest",
        },
        "emoticon": "",
        "enlargeEmoji": false,
        "hexcode": "1F408",
        "renderUnicode": false,
        "shortcode": "",
        "unicode": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "1F408",
      "nodeType": "host",
      "props": Object {
        "aria-label": "cat",
        "children": <Emoji
          emojiLargeSize="3em"
          emojiPath="{{hexcode}}"
          emojiSize={22}
          emojiSource={
            Object {
              "compact": false,
              "locale": "en",
              "version": "latest",
            }
          }
          emoticon=""
          enlargeEmoji={false}
          hexcode="1F408"
          renderUnicode={false}
          shortcode=""
          unicode=""
        />,
        "className": "interweave-picker__emoji",
        "onClick": [Function],
        "onMouseEnter": [Function],
        "onMouseLeave": [Function],
        "style": Object {
          "height": 32,
          "padding": 5,
          "width": 32,
        },
        "type": "button",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "emojiLargeSize": "3em",
          "emojiPath": "{{hexcode}}",
          "emojiSize": 22,
          "emojiSource": Object {
            "compact": false,
            "locale": "en",
            "version": "latest",
          },
          "emoticon": "",
          "enlargeEmoji": false,
          "hexcode": "1F408",
          "renderUnicode": false,
          "shortcode": "",
          "unicode": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
